src/algorithms/signal_source/libs/ion_gsms.cc

```diff
@@ -0,0 +1,159 @@
+ * This file is part of GNSS-SDR.
+ *
+ * Copyright (C) 2010-2020  (see AUTHORS file for a list of contributors)
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ * -----------------------------------------------------------------------------
+ */
+
+#include "ion_gsms.h"
+#include "gnuradio/block.h"
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#if USE_GLOG_AND_GFLAGS
+#include <glog/logging.h>
+#else
+#include <absl/log/log.h>
+#include <utility>
+#endif
+
+using namespace std::string_literals;
+
+IONGSMSFileSource::IONGSMSFileSource(
+    const ConfigurationInterface* configuration __attribute__((unused)),
+    const std::string& role __attribute__((unused)),
+    const std::filesystem::path& metadata_filepath,
+    const GnssMetadata::File& file,
+    const GnssMetadata::Block& block,
+    const std::vector<std::string>& stream_ids)
+    : gr::sync_block(
+          "ion_gsms_file_source",
+          gr::io_signature::make(0, 0, 0),
+          make_output_signature(block, stream_ids)),
+      file_metadata_(file),
+      block_metadata_(block),
+      io_buffer_offset_(0),
+      maximum_item_rate_(0),
+      chunk_cycle_length_(0)
+{
+    std::filesystem::path data_filepath = metadata_filepath.parent_path() / file.Url().Value();
+    fd_ = std::fopen(data_filepath.c_str(), "rb");
+    std::size_t block_offset = file.Offset();
+    std::fseek(fd_, file.Offset() + block_offset + block.SizeHeader(), SEEK_SET);
+
+    std::size_t output_stream_offset = 0;
+    for (const auto& chunk : block.Chunks())
+        {
+            chunk_data_.emplace_back(std::make_shared<IONGSMSChunkData>(chunk, stream_ids, output_stream_offset));
+            chunk_cycle_length_ += chunk.CountWords() * chunk.SizeWord();
+            const std::size_t out_count = chunk_data_.back()->output_stream_count();
+            output_stream_offset += out_count;
+            for (std::size_t i = 0; i < out_count; ++i)
+                {
+                    output_stream_item_sizes_.push_back(chunk_data_.back()->output_stream_item_size(i));
+                    maximum_item_rate_ = std::max(chunk_data_.back()->output_stream_item_rate(i), maximum_item_rate_);
+                }
+        }
+    output_stream_count_ = output_stream_offset;
+}
+
+IONGSMSFileSource::~IONGSMSFileSource()
+{
+    std::fclose(fd_);
+}
+
+int IONGSMSFileSource::work(
+    int noutput_items,
+    gr_vector_const_void_star& input_items __attribute__((unused)),
+    gr_vector_void_star& output_items)
+{
+    const std::size_t max_sample_output = std::floor((noutput_items - 1.0) / maximum_item_rate_);
+    io_buffer_.resize(max_sample_output * chunk_cycle_length_);
+    io_buffer_offset_ = 0;
+    std::fread(io_buffer_.data(), sizeof(decltype(io_buffer_)::value_type), io_buffer_.size(), fd_);
+
+    items_produced_.resize(output_items.size());
+    for (int& i : items_produced_)
+        {
+            i = 0;
+        }
```

Comment:
This seems a bit obtuse. Consider:
```suggestion
    items_produced_.clear();
    items_produced_.resize(output_items.size());
```
Note that `resize()` initializes all _new_ items to `T()` so by making the vector length 0 before resizing it, you ensure all the valuse are default initialized.

Reply:
Fixed